src/openzeppelin/token/erc1155/library.cairo

```diff
@@ -0,0 +1,601 @@
+// SPDX-License-Identifier: MIT
+// OpenZeppelin Contracts for Cairo v0.5.2 (token/erc1155/library.cairo)
+
+%lang starknet
+
+from starkware.starknet.common.syscalls import get_caller_address
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.cairo.common.math import assert_not_zero, assert_not_equal
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.uint256 import Uint256, uint256_check
+from starkware.cairo.common.bool import TRUE
+
+from openzeppelin.introspection.erc165.IERC165 import IERC165
+from openzeppelin.introspection.erc165.library import ERC165
+from openzeppelin.token.erc1155.IERC1155Receiver import IERC1155Receiver
+from openzeppelin.security.safemath.library import SafeUint256
+from openzeppelin.utils.constants.library import (
+    IERC1155_ID,
+    IERC1155_METADATA_ID,
+    IERC1155_RECEIVER_ID,
+    IACCOUNT_ID,
+    ON_ERC1155_RECEIVED_SELECTOR,
+    ON_ERC1155_BATCH_RECEIVED_SELECTOR,
+)
+
+//
+// Events
+//
+
+@event
+func TransferSingle(operator: felt, from_: felt, to: felt, id: Uint256, value: Uint256) {
+}
+
+@event
+func TransferBatch(
+    operator: felt,
+    from_: felt,
+    to: felt,
+    ids_len: felt,
+    ids: Uint256*,
+    values_len: felt,
+    values: Uint256*,
+) {
+}
+
+@event
+func ApprovalForAll(account: felt, operator: felt, approved: felt) {
+}
+
+@event
+func URI(value: felt, id: Uint256) {
+}
+
+//
+// Storage
+//
+
+@storage_var
+func ERC1155_balances(id: Uint256, account: felt) -> (balance: Uint256) {
+}
+
+@storage_var
+func ERC1155_operator_approvals(account: felt, operator: felt) -> (approved: felt) {
+}
+
+@storage_var
+func ERC1155_uri() -> (uri: felt) {
+}
+
+namespace ERC1155 {
+    //
+    // Initializer
+    //
+
+    func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(uri: felt) {
+        _set_uri(uri);
+        ERC165.register_interface(IERC1155_ID);
+        ERC165.register_interface(IERC1155_METADATA_ID);
+        return ();
+    }
+
+    //
+    // Modifiers
+    //
+
+    func assert_owner_or_approved{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        owner
+    ) {
+        let (caller) = get_caller_address();
+        if (caller == owner) {
+            return ();
+        }
+        let (approved) = ERC1155.is_approved_for_all(owner, caller);
+        with_attr error_message("ERC1155: caller is not owner nor approved") {
+            assert approved = TRUE;
+        }
+        return ();
+    }
+
+    //
+    // Getters
+    //
+
+    func uri{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(id: Uint256) -> (
```

Comment:
```suggestion
    // This implementation returns the same URI for *all* token types. It relies
    // on the token type ID substitution mechanism
    // https://eips.ethereum.org/EIPS/eip-1155#metadata[defined in the EIP].
    //
    // Clients calling this function must replace the `\{id\}` substring with the
    // actual token type ID.
    func uri{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(id: Uint256) -> (
```

I think comments like this are important enough to be not only in the docs (where I assume it will be added) but also in the source code.

Is hard to understand how to use this URI without this notice.

Maybe it make more sense in the preset? I would add comments on both. Maybe not the same, but a link from the preset comment to the library comment.

Reply:
I think here in the library is enough, which is what the user should read to understand what's happening